### ISSUE TITLE
Support multiple races of characters

### ISSUE THREAD

**Issue by kagekiyo7** [MotocalDevelopers/motocal#296]
![image](https://user-images.githubusercontent.com/22173829/61121871-48f01f00-a4db-11e9-9055-9988a8dcdcff.png)

**Comment by Groovman** [comment]
either we add a 2nd race field which will probablz break data loading, or we add concatinated multi races and check for startswith/endswith/substring for the race instead of equal.

**Comment by kei-gbf** [comment]
@Groovman 

String.includes
- `"human/erune".includes("human")` => true
- actual code: `race === "human" || race.includes("human")`
  - the most of characters does not have 2nd race, the first check enough.
  - maybe not so simple, in baha weapon case it contains seisyo and unknown

I did not research the performance, browser support the String API etc.
just info for choice now.

I have more topics about the race,element,fav checking, I've worked on #259
(not merged to master yet) It aim to support "range" fields:
"range": "doraf|seisyo|unknown" in consts. (not implemented yet)

----
- favorite weapons is multiple field already they use two fields "fav1", "fav2"
- we need discussion
  - it should be "race", "race2" ?
  - or extend single field "race": "human/erune"

if we check race by those String methods, extend the current single field is looks better.

----
we have hardcoded the race checking but, it was not so many places as I thought.

- arm ... weapon skill consts does not have race info, it's coded in logic code directly.
  - haba weapons
  - seisyo
    - [ ] "seisyo" may not contain "unknown" ?
- support
  - doraf/unknown/seisyo

I can work for the race checking function, if you could provide multiple race in json

**Comment by kei-gbf** [comment]
sample code + unit tests for the race check function
https://codepen.io/anon/pen/ydrZQX?editors=0010

**Comment by Groovman** [comment]
Seisyo is Primal, unknowns don't get specific primal effects. Baha applies to unknown and primals, probably because of balance reasons. Basically what is affect is decided by cygames at random.

I would favor a single race field because most logic would go in the python script and the application would just need to check for String.include() (I would leave the equal check out just to make code smaller), even if cygames decides to do 3 races characters the app code would not need to be touched. While seperate race fields would require app code changes.

I like the function you can pass an array to because it makes code much more readable than having to read a sequences of boolean checks.

**Comment by kei-gbf** [comment]
@Groovman 
I got questions and confirms

----
### support: "normalBuff_doraf"

```js
// Draph, Unknown and Primal characters only (Fortified Vigor)
if (totals[key2]["race"] === "doraf" ||
    totals[key2]["race"] === "unknown" ||
    totals[key2]["race"] === "seisho") {
    totals[key2]["normalBuff"] += support.value
}
```

- [ ] No "wildcard" check for support? or can I add the wildcard check here?
  - Kaimu support text describes it's for weapon skill, not mention to support.
     but actual verification? if game internally use the same logic for race checking.

----
### Baha race check included flags

following code, it's refactored. (currently, in my local work)
remove clone codes, wildcard check is in the function.
additionally, it will check support,support2,support3 fields too.

```js
if (stype == 'bahaAT' && !isBahaAtIncluded && bahaRaceCharaContains(skillname, totals[key])) {
    // Baha dagger etc.
    totals[key]["bahaAT"] += comb[i] * skillAmounts["bahaAT"][amount][slv - 1];
    isBahaAtIncluded = true;
} else if (stype == 'bahaATHP' && !isBahaAthpIncluded && bahaRaceCharaContains(skillname, totals[key])) {
    // Baha sword etc.
    totals[key]["bahaAT"] += comb[i] * skillAmounts["bahaAT"][amount][slv - 1];
    totals[key]["bahaHP"] += comb[i] * skillAmounts["bahaHP"][amount][slv - 1];
    isBahaAthpIncluded = true;
} else if (stype == 'bahaHP' && !isBahaHpIncluded && bahaRaceCharaContains(skillname, totals[key])) {
    // Baha Fist etc
    totals[key]["bahaHP"] += comb[i] * skillAmounts["bahaHP"][amount][slv - 1];
    isBahaHpIncluded = true;
} else if (stype == 'bahaFUATHP' && bahaFURaceCharaContains(skillname, totals[key])) {
    totals[key]["bahaAT"] += comb[i] * skillAmounts["bahaFUATHP"]["AT"][slv - 1];
    totals[key]["bahaHP"] += comb[i] * skillAmounts["bahaFUATHP"]["HP"][slv - 1];
} else if (stype == 'bahaFUHP' && bahaFURaceCharaContains(skillname, totals[key])) {
    totals[key]["bahaHP"] += comb[i] * skillAmounts["bahaFUHP"]["HP"][slv - 1];
    totals[key]["bahaDA"] += comb[i] * skillAmounts["bahaFUHP"]["DA"][slv - 1];
    totals[key]["bahaTA"] += comb[i] * skillAmounts["bahaFUHP"]["TA"][slv - 1];
```

- [ ] no need included flags for bahaFU ? (confirm)
- [ ] I wonder how the included flag work when two baha weapon same stype but different race.

----
### wildcard check in Cosmos weapons

```js
// Cosmos weapons
if ((skillname == 'cosmosAT' && totals[key]["type"] == "attack") || totals[key]["support"] === "wildcard")
```

- [ ] seem ignore skillname check when chara has wildcard support.
  - [ ] in the chain of if/else, wildcard always get cosmosAT affect?

**Comment by kei-gbf** [comment]
- [ ] Can I add "wildcard" check for "omega-gekijou" and "omega-kyousou" ? (in generateHaisuiData)

**Comment by Groovman** [comment]
> Subject to all specialty weapon-, style-, and race-related weapon skills

From the text I assume it does not affect other character support skills only skills from weapons. "omega-gekijou" and "omega-kyousou" should get the wildcard check.

**Comment by Groovman** [comment]
Closed with #302 being finsihed.
